src/observer/ingest_logic.py

```diff
@@ -416,21 +418,61 @@ def mkidx():
 def _download_versions(msid, latest_version):
     LOG.info(' %s versions to fetch' % latest_version)
     version_range = range(1, latest_version + 1)
-    lmap(lambda version: upsert_ajson(consume("articles/%s/versions/%s" % (msid, version))), version_range)
+
+    def fetch(version):
+        upsert_ajson(msid, version, models.LAX_AJSON, consume("articles/%s/versions/%s" % (msid, version)))
+    lmap(fetch, version_range)
 
 def download_article_versions(msid):
-    "loads *all* versions of given article from the api"
+    "loads *all* versions of given article via API"
     resp = consume("articles/%s/versions" % msid)
-    return _download_versions(msid, len(resp["versions"]))
+    _download_versions(msid, len(resp["versions"]))
 
 def download_all_article_versions():
-    "loads *all* versions of *all* articles from the api"
-    msid_ver_idx = mkidx()
+    "loads *all* versions of *all* articles via API"
+    msid_ver_idx = mkidx() # urgh. this sucks. lax needs a /summary endpoint too
```

Comment:
in the sense that we need to staple together all `/articles/:id/versions` responses (paginated by N articles at a time)? Or even more like all `/articles/:id/versions/:v` responses which are not snippets?

It looks like a large response. If Observer needs all that data, it can make local copies and update each upon the `bus` event. That way it always has all the latest versions without creating load on Lax to generate summaries (which should be the job of Observer in fact).

Reply:
the first case - we need to scrape every version of every article. snippet isn't necessary.

observer has two mechanisms for updating, the article update listener which does articles piecemeal as it receives an event, and a full scrape, which is part of what we're looking at here. when the update listener has been down or encounters an error  or we introduce new data that needs to be scraped (like this pr), a full scrape is done and the results are stored in the database.

in order to do this full scrape we need to know which articles exist and their latest version. Observer hits the article history endpoint until it has the msid and max version of every article, the rest of the snippet is thrown away. it then does many thousands of requests to lax to get the full data.

what the comment is complaining about is that I don't need the snippet that comes with traversing the article history endpoint. if lax had a /summary endpoint I could get a good idea of what articles existed and their versions more quickly. same number of requests, just a much much smaller response.